src/libraries/System.Private.CoreLib/src/System/Buffers/Binary/BinaryPrimitives.ReverseEndianness.cs

```diff
@@ -0,0 +1,260 @@
+                !source.Overlaps(destination, out int elementOffset) ||
+                elementOffset < 0)
+            {
+                // Either there's no overlap between the source and the destination, or there's overlap but the
+                // destination starts at or before the source.  That means we can safely iterate from beginning
+                // to end of the source and not have to worry about writing into the destination and clobbering
+                // source data we haven't yet read.
+
+                int i = 0;
+
+                if (Vector256.IsHardwareAccelerated)
+                {
+                    while (i <= source.Length - Vector256<T>.Count)
+                    {
+                        Vector256.StoreUnsafe(TReverser.Reverse(Vector256.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                        i += Vector256<T>.Count;
+                    }
+                }
+
+                if (Vector128.IsHardwareAccelerated)
+                {
+                    while (i <= source.Length - Vector128<T>.Count)
+                    {
+                        Vector128.StoreUnsafe(TReverser.Reverse(Vector128.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                        i += Vector128<T>.Count;
+                    }
+                }
+
+                while (i < source.Length)
+                {
+                    Unsafe.Add(ref destRef, i) = TReverser.Reverse(Unsafe.Add(ref sourceRef, i));
+                    i++;
+                }
+            }
+            else
+            {
+                // There's overlap between the source and the destination, and the source starts before the destination.
+                // That means if we were to iterate from beginning to end, reading from the source and writing to the
+                // destination, we'd overwrite source elements not yet read.  To avoid that, we iterate from end to beginning.
+
+                int i = source.Length;
+
+                if (Vector256.IsHardwareAccelerated)
+                {
+                    while (i >= Vector256<T>.Count)
+                    {
+                        i -= Vector256<T>.Count;
+                        Vector256.StoreUnsafe(TReverser.Reverse(Vector256.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                    }
+                }
+
+                if (Vector128.IsHardwareAccelerated)
+                {
+                    while (i >= Vector128<T>.Count)
+                    {
+                        i -= Vector128<T>.Count;
+                        Vector128.StoreUnsafe(TReverser.Reverse(Vector128.LoadUnsafe(ref sourceRef, (uint)i)), ref destRef, (uint)i);
+                    }
+                }
+
+                while (i > 0)
+                {
+                    i--;
+                    Unsafe.Add(ref destRef, i) = TReverser.Reverse(Unsafe.Add(ref sourceRef, i));
+                }
+            }
```

Comment:
Have you benchmarked the difference between iterating from the start vs from the end? I wonder if we should just always iterate from the end?

Reply:
We can't always iterate from the end as long as there's the possibility of non-perfect overlap.  Let's say you have:
```C#
Span<char> orig = new char[] { 'a', 'b', 'c', 'd', 'e' };
Span<char> src = orig.Slice(2, 3);
Span<char> dst = src.Slice(0, 3);
Copy(src, dst);
```
To keep things simply I'll use a Copy instead of ReverseEndianness (which is just a copy that manipulates each value).  The resulting `orig` should be:
```
'c', 'd', 'e', 'd', 'e'
```
but if we copied left to right, we'd get:
```
'e', 'd', 'e', 'd', 'e'
```
because you'd start by copying src[2] (which is orig[4]) to dst[2] (which is orig[2]), overwriting the 'c' that we still haven't copied yet.

Reply:
Thanks for a detailed answer!